## Gather

In [1]:
#Import the neccessary libraries

import pandas as pd
import numpy as np

In [2]:
#Loading Data sets
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

> Assesing the data is the second step in data wrangling after gathering data, were you explore your data to have a preif understanding of the datasets structures, how to deal with their data, and figure out the problems they might have and configre some solutions while exploring ..

In [3]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [4]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [5]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
#Viewing the data columns names
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex given_name       surname                 address  \
237         238         male       John           Doe         123 Main Street   
205         206         male      Breno       Correia        3539 Bottom Lane   
181         182       female     Yumena      Nakayama  4458 Stark Hollow Road   
89           90         male      Kwasi         Mensa      479 Elmwood Avenue   
21           22         male     Samúel  Guðbrandsson     1904 Granville Lane   

          city     state  zip_code        country  \
237   New York        NY   12345.0  United States   
205  Tonawanda  New York   14150.0  United States   
181     Gypsum        CO   81637.0  United States   
89       Tempe        AZ   85281.0  United States   
21    Elmsford        NJ   10523.0  United States   

                                        contact  birthdate  weight  height  \
237                 johndoe@email.com1234567890   1/1/1975   180.0      72   
205  716-743-5884BrenoLimaCorreia@superrito.com  7/11/1974   180.6      67   
181      YumenaNakayama@gustr.com1 970 754 5843  3/24/1929   130.2      59   
89         KwasiMensa@cuvox.de+1 (480) 967-6336  10/5/1942   145.4      69   
21    973-445-5341SamuelGubrandsson@teleworm.us  4/12/1983   223.7      69   

      bmi  
237  24.4  
205  28.3  
181  26.3  
89   21.5  
21   33.0

In [15]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Schiavone      2
Woźniak        2
              ..
Petersen       1
Haraguchi      1
Koivuniemi     1
Olofsson       1
Eldarkhanov    1
Name: surname, Length: 466, dtype: int64

In [16]:
patients.address.value_counts()

123 Main Street           6
648 Old Dear Lane         2
2476 Fulton Street        2
2778 North Avenue         2
1733 Blackwell Street     1
                         ..
720 Tator Patch Road      1
4324 Oak Street           1
4571 Walnut Hill Drive    1
1703 Everette Alley       1
2625 Romines Mill Road    1
Name: address, Length: 483, dtype: int64

In [17]:
#Check duplicates for the address column
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [19]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [20]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality
-- Here are multiple quality problems found during assessing step in the provided datasets 
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
-- Here are multiple tidiness problems found during assessing step in the provided datasets 

- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [23]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

#### `treatments`: Missing records (280 instead of 350)

##### Define
-Import the missing records from the reatments_cut.csv

-concate dataframes together (Outputting one full dataframe)
<br>             

</br>

* Note: the missing `treatments` records are stored in a file named `treatments_cut.csv`*

##### Code

In [24]:
# Your cleaning code here
treat_miss = pd.read_csv('treatments_cut.csv')
treatments_clean = pd.concat([treatments_clean ,treat_miss], ignore_index = True)

##### Test

In [25]:
# Your testing code here
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [26]:
treatments_clean.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)


##### Define
-Recalculate the `hba1c_change` column using the formula : `hba1c_start` - `hba1c_end`

##### Code

In [27]:
# Your cleaning code here
treatments_clean['hba1c_change'] = treatments_clean['hba1c_start'] - treatments_clean['hba1c_end']

##### Test

In [28]:
# Your testing code here
treatments_clean['hba1c_change'].head()

0    0.43
1    0.47
2    0.43
3    0.35
4    0.32
Name: hba1c_change, dtype: float64

### Tidiness

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
> Extract the phone number and email from `contact` column with str.extract method that uses regular expression, then remove the `contact` column when done.



##### Code

In [29]:
# Your cleaning code here

#Adding the phoneNumber column to the data frame to extract the phone from contact column
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)
#Adding the Email column to the data frame to extract the email address from contact column
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

#Drop contact column (since replaced with 2 extra columns phone,mail )
patients_clean = patients_clean.drop('contact', axis=1)

##### Test

In [30]:
# Your testing code here

#You can use .head() with 0 sent in parameter to show the outline of the table
patients_clean.head(0)

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [31]:
#Or you can only list the columns names [which is prefered from my side (; ]
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [32]:
#Checking a small sample of the phone number to ensure the correction of the process
patients_clean.phone_number.sample(5)

262         713-556-4328
501         360 443 2060
184    +1 (386) 989-0019
139    +1 (513) 383-0516
446         601-418-0102
Name: phone_number, dtype: object

In [33]:
#Checking a small sample of the mail addresses to ensure the correction of the process
patients_clean.email.sample(5)

147          KristianVukelic@teleworm.us
125    SewerynaCzerwinska@jourrapide.com
471       NadwahHawadahNaifeh@einrot.com
345            AsunaMorita@superrito.com
399                EricEk@jourrapide.com
Name: email, dtype: object

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define

*Hint: use pandas' [melt function](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html) and [`str.split()` method](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.split.html)*
> Melt the two columns `auralin` and `novodra` to one column `treatment`, then store it's values into another column called `dose` which wil have the full representation same as before, then split this column using str.split method into start and end dose by slicing it from the ` - `, lastly remove the `dose` column.

##### Code

In [34]:
# Your cleaning code here

#Adding the melt function 
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],var_name='treatment', value_name='dose')
#Add only the rows that contains a value, if = '-' means it doesn't have a treatment, so not added 
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
#Splitting the 'dose' column into start and end dose 
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str

#Removing the 'dose' column after splitting its values in the previous step
treatments_clean = treatments_clean.drop('dose', axis=1)

<ipython-input-34-dad9f28030d5>:8: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str


##### Test

In [35]:
# Your testing code here
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

#### Adverse reaction should be part of the `treatments` table

##### Define
>Merge `adverse_reaction` column into treatments table, joining on `given name` and `surname`.




*Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html) for the function used in the solution.*

##### Code

In [36]:
# Your cleaning code here
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on = ['given_name', 'surname'], how = 'left')

##### Test

In [37]:
# Your testing code here
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
1       skye  gormanston         7.97       7.62          0.35   auralin   
2     sophia      haugen         7.65       7.27          0.38   auralin   
3      eddie      archer         7.89       7.55          0.34   auralin   
4       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end adverse_reaction  
0        41u      48u              NaN  
1        33u      36u              NaN  
2        37u      42u              NaN  
3        31u      38u              NaN  
4        30u      36u              NaN

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
Hint: [tutorial](https://chrisalbon.com/python/pandas_join_merge_dataframe.html)  and [tutorial](http://erikrood.com/Python_References/dropping_rows_cols_pandas.html)

>Split the the `patient_id` and `name` in the `patients` table, then convert these names to lower case to join with `treatments` names. Then remove the two `name`and `surname` columns in the treatments table.

##### Code

In [38]:
# Your cleaning code here

id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()#Make sure all strings are lower case 
id_names.surname = id_names.surname.str.lower()#Make sure all strings are lower case 
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])

#Removing given_name and surname columns
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

M:\Program files\Anaconda\anacondafile\lib\site-packages\pandas\core\generic.py:5168: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[name] = value


##### Test

In [39]:
# Your testing code here

#To check if the process done successfully  
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0         7.63       7.20          0.43   auralin        41u      48u   
1         7.97       7.62          0.35   auralin        33u      36u   
2         7.65       7.27          0.38   auralin        37u      42u   
3         7.89       7.55          0.34   auralin        31u      38u   
4         7.76       7.37          0.39   auralin        30u      36u   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

In [40]:
#Here the 'patient_id' is going to be the only duplicate that should appear
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

### Quality

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define


> Changing the zip code column data type from float to string using `astype`

> then remove the '.0' using string slicing

> and replace the 0000n values that were converted because of the type conversion into NaN for digit zip codes .

##### Code

In [41]:
# Your cleaning code here

patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')
#Replace values
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

##### Test

In [42]:
# Your testing code here
patients_clean['zip_code'].head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
*Your definition here.*

>Replace height for rows in the `patients` table that have a height of 27 in (there is only one) with 72 in.

##### Code

In [43]:
# Your cleaning code here
#Direct change since it's only one value
patients_clean.height = patients_clean.height.replace(27, 72)

##### Test

In [44]:
# Your testing code here

# Should be empty since there is no such height
patients_clean[patients_clean.height == 27]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [45]:
# Confirm the replacement worked
patients_clean[patients_clean.surname == 'Neudorf']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

   phone_number                email  
4  334-515-7487  TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
Hint: [tutorial](https://chrisalbon.com/python/pandas_apply_operations_to_dataframes.html) 

>Adding a function to convert the full state name to state abbreviation for ( California, New York, Illinois, Florida, and Nebraska).

##### Code

In [46]:
# Your cleaning code here

#linking each state with it's abbreviation
state_abb = {'California': 'CA','New York': 'NY','Illinois': 'IL','Florida': 'FL','Nebraska': 'NE'}


def abb_state(patient):
    if patient['state'] in state_abb.keys():
        abb = state_abb[patient['state']]
        return abb
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abb_state, axis=1)

##### Test

In [47]:
# Your testing code here

patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
FL    22
MA    22
PA    18
GA    15
OH    14
MI    13
LA    13
OK    13
NJ    12
VA    11
MS    10
WI    10
IN     9
MN     9
AL     9
TN     9
WA     8
KY     8
NC     8
MO     7
NE     6
NV     6
KS     6
ID     6
CT     5
SC     5
IA     5
ME     4
AR     4
CO     4
ND     4
AZ     4
RI     4
MD     3
DE     3
OR     3
WV     3
SD     3
MT     2
DC     2
VT     2
AK     1
NM     1
NH     1
WY     1
Name: state, dtype: int64

#### Dsvid Gustafsson

##### Define


>Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'.

##### Code

In [48]:
# Your cleaning code here

#Direct change
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### Test

In [49]:
# Your testing code here

patients_clean[patients_clean.surname == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
Hint: [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.astype.html) , [documentation page](http://pandas.pydata.org/pandas-docs/version/0.20/generated/pandas.to_datetime.html) , and [documentation page](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.strip.html) 

>Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. Convert birthdate to datetime data type. Cut and remove the letter 'u' in start and end dose then convert them into data type integer.


##### Code

In [50]:
# Your cleaning code here
#Category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

#Datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

#Cut u using strip function and changing data type to integer
treatments_clean.dose_start = treatments_clean.dose_start.str.strip('u').astype(int)
treatments_clean.dose_end = treatments_clean.dose_end.str.strip('u').astype(int)

##### Test

In [51]:
# Your testing code here

patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    category      
 7   zip_code      491 non-null    object        
 8   country       491 non-null    object        
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int64         
 12  bmi           503 non-null    float64       
 13  phone_number  491 non-null    object        
 14  email         491 non-null    object        
dtypes: category(2), datetime64[ns](1), float

In [52]:
#Checking the start and endd dose data type if successfully converted
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   hba1c_start       349 non-null    float64
 1   hba1c_end         349 non-null    float64
 2   hba1c_change      349 non-null    float64
 3   treatment         349 non-null    object 
 4   dose_start        349 non-null    int32  
 5   dose_end          349 non-null    int32  
 6   adverse_reaction  35 non-null     object 
 7   patient_id        349 non-null    int64  
dtypes: float64(3), int32(2), int64(1), object(2)
memory usage: 21.8+ KB


#### Multiple phone number formats

##### Define
Hint: helpful [Stack Overflow answer](https://stackoverflow.com/a/123681).

>Cut all " ", "-", "(", ")", and "+" with and store each number without any formatting. Pad the phone number with a 1 if the length of the number is 10 digits.

##### Code

In [53]:
# Your cleaning code here

patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [54]:
# Your testing code here

patients_clean.phone_number.head()

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

#### Default John Doe data

##### Define


>Remove John Doe records from the `patients` table, since as assumed they are non-recoverable..

##### Code

In [55]:
# Your cleaning code here

patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [56]:
# Your testing code here

# Should be no Doe records
patients_clean.surname.value_counts()

Jakobsen       3
Taylor         3
Lương          2
Collins        2
Woźniak        2
              ..
Petersen       1
Haraguchi      1
Koivuniemi     1
Olofsson       1
Eldarkhanov    1
Name: surname, Length: 465, dtype: int64

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
helpful: [boolean indexing](http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing)

>Remove (Jake Jakobsen, Pat Gersten, and Sandy Taylor) rows from the `patients` table. These are the nicknames that are not in the `treatments` table .

##### Code

In [57]:
# Your cleaning code here


patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [58]:
# Your testing code here

#Checking each surname to search for the removed names (shoudn't be appearing)
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [59]:
patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

   phone_number                     email  
97  14028484923  PatrickGersten@rhyta.com

In [60]:
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

    phone_number                       email  
131  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

#### kgs instead of lbs for Zaitseva weight

##### Define
Hint :[advanced indexing](https://stackoverflow.com/a/44913631)

>Isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

##### Code

In [61]:
# Your cleaning code here

#Since the row we want contains the lowest wight, we search using .min() function
weight_kg = patients_clean.weight.min()
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
#Recalculate the wight (from kg to lb)
patients_clean.loc[mask, column_name] = weight_kg * 2.20462

##### Test

In [62]:
# Your testing code here

# 48.8 shouldn't be the lowest anymore
patients_clean.weight.sort_values()

459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 494, dtype: float64